arrows/gdal/image_io.cxx

```diff
@@ -0,0 +1,277 @@
+    GDALOpen( filename.c_str(), GA_ReadOnly ) ) );
+
+  vital::image img;
+
+  if ( !gdalDataset )
+  {
+    VITAL_THROW( vital::invalid_file, filename, "GDAL could not load file.");
+  }
+  else
+  {
+    // Get the size of the image and the number of bands
+    auto imgWidth = gdalDataset->GetRasterXSize();
+    auto imgHeight = gdalDataset->GetRasterYSize();
+    auto imgDepth = gdalDataset->GetRasterCount();
+
+    // Create a new vital image based on the GDAL raster type. For now just
+    // load bands directly into channels in vital::image.
+    // TODO: deal or provide warning if bands have different types.
+    auto bandType = gdalDataset->GetRasterBand(1)->GetRasterDataType();
+    switch (bandType)
+    {
+      case (GDT_Byte):
+      {
+        img = vital::image_of<vital::byte>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_UInt16):
+      {
+        img = vital::image_of<uint16_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Int16):
+      {
+        img = vital::image_of<int16_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_UInt32):
+      {
+        img = vital::image_of<uint32_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Int32):
+      {
+        img = vital::image_of<int32_t>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Float32):
+      {
+        img = vital::image_of<float>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      case (GDT_Float64):
+      {
+        img = vital::image_of<double>(imgWidth, imgHeight, imgDepth);
+        break;
+      }
+      default:
+      {
+        std::stringstream ss;
+        ss << "kwiver::arrows::gdal::image_io::load(): "
+           << "Unknown or unsupported pixal type: "
+           << GDALGetDataTypeName(bandType);
+        VITAL_THROW( vital::image_type_mismatch_exception, ss.str() );
+        break;
+      }
+    }
+
+    // Loop over bands and copy data
+    CPLErr err;
+    for (int i = 1; i <= imgDepth; ++i)
+    {
+      GDALRasterBand* band = gdalDataset->GetRasterBand(i);
+      bandType = band->GetRasterDataType();
+      err = band->RasterIO(GF_Read, 0, 0, imgWidth, imgHeight,
+        static_cast<void*>(reinterpret_cast<GByte*>(
+          img.first_pixel()) + (i-1)*img.d_step()*img.pixel_traits().num_bytes),
+        imgWidth, imgHeight, bandType, 0, 0);
+    }
+
+    // Get and translate metadata
+    vital::metadata_sptr md = std::make_shared<vital::metadata>();
+
+    // Get geotransform and calculate corner points
+    double geo_transform[6];
+    gdalDataset->GetGeoTransform(geo_transform);
+
+    OGRSpatialReference osrs;
+    auto proj_ptr = gdalDataset->GetProjectionRef();
+    char* wktDesc = new char[strlen(proj_ptr)];
+    strcpy( wktDesc, proj_ptr ); // Copy needed due const/non-const api conflicts
+    char* tmpPtr = wktDesc; // This gets advanced by importFromWkt
+    osrs.importFromWkt( &tmpPtr );
+
```

Comment:
I know the preceding block of code is a mess but `GetProjectionRef()` returns a `const char *` and `importFromWkt()` takes a `char **`. This could be simplified if we moved to GDAL 2.2.3 or greater. I am open to any suggestions for a way around this.